### DISCUSSION TITLE
DC through windows OS for experimental lab

### DISCUSSION THREAD

**Discussion by giper45** [copy/v86#1195] [Q&A] [answered]
First of all, thank you for this awesome project!
I would like to know if, according to your expertise, it would be feasible to setup a Domain Controller for Windows through your lightweight Win machines.
I am trying to develop a security playground in WASM, in the same way I used to do through [Docker](https://github.com/DockerSecurityPlayground/DSP), and this project is really one of the most interesting things I have looked to until now.
Even if you could think that my request could be `insane`, or impossible, thank you very much your contribution in the WebAssembly ecosystem world!

**Comment by copy** [comment]
A good starting point would be to create an XP or 2000 image in qemu, and play around with it in v86: https://github.com/copy/v86/blob/master/docs/windows-xp.md

Networking is somewhat messy and under-documented (suggestions welcome), but your options are roughly:

- Running a relay server, see https://github.com/copy/v86/issues/1094 and regarding performance, https://github.com/copy/v86/discussions/1175#discussioncomment-11203479
- Hook up multiple v86 instances on the same local machine with some JavaScript, e.g. https://github.com/copy/v86/blob/master/examples/two_instances.html or https://github.com/copy/v86/issues/1094 (you will need to configure static ip addresses)

**Reply by SuperMaxusa** [reply]
In addition, I would comment these options:
 - Relay servers (such as [websockproxy](https://github.com/benjamincburns/websockproxy) and [node-relay](https://github.com/krishenriksen/node-relay)) have built-in DHCP and DNS servers (if it won't interfere for Domain Controller), and you can also have to access to the Internet, local networking works as expected: https://github.com/copy/v86/issues/783#issuecomment-1344767981. 
 Note that this applies only to WebSocket servers with tuntap, it didn't work for wisp and fetch as they use internal DHCP and DNS stacks. 
 - The second option should suit your tasks, you may not have access to the Internet, only within the local network. To avoid using static IPs, you need to run DHCP and DNS server on your Domain Controller server.

**Comment by giper45** [comment, marked as answer]
And ... the domain join occurred!
<img width="838" alt="Screenshot 2024-12-14 at 17 08 03" src="https://github.com/user-attachments/assets/93b6d641-3764-461e-8be2-992f5bb5da04" />

<img width="593" alt="image" src="https://github.com/user-attachments/assets/60cc549b-ed9a-4369-b77c-6616249046f5" />


Followed steps:
* Create images (winxp and win2k3 using the [XP guide](https://github.com/copy/v86/blob/master/docs/windows-xp.md)
* Configure on QEMU the DC on win2k3 server using the following [guide](https://www.visualwin.com/AD-Controller/)
* Created a `dsp` user in the domain controller
* Modified [two instances](https://github.com/copy/v86/blob/master/examples/two_instances.html) example to configure the domain and run the two images on the browser
* Join the XP instance using the user `dsp` (steps [here](https://www.visualwin.com/AD-XP/)) 

The user profile is loaded in about 5 minutes, but everything works. No Internet connection (I was interested in realizing the domain network)

I really like it, even if it is not useful... it is just very funny for me realizing a DC in the browser!!!

Thank you for your suggestions, and the project, I will be happy to contribute in my spare time.

**Reply by Darin755** [reply]
That's pretty cool
